File: 31. Next Permutation/31. Next Permutation.js
function nextPermutation(nums) {
    let n = nums.length;
    
    // Step 1: Find the largest index k such that nums[k] < nums[k + 1]
    let k = n - 2;
    while (k >= 0 && nums[k] >= nums[k + 1]) {
        k--;
    }
    
    if (k >= 0) {
        // Step 2: Find the largest index l greater than k such that nums[k] < nums[l]
        let l = n - 1;
        while (nums[l] <= nums[k]) {
            l--;
        }
        
        // Step 3: Swap nums[k] and nums[l]
        [nums[k], nums[l]] = [nums[l], nums[k]];
    }
    
    // Step 4: Reverse the sequence from k + 1 to the end
    let start = k + 1;
    let end = n - 1;
    while (start < end) {
        [nums[start], nums[end]] = [nums[end], nums[start]];
        start++;
        end--;
    }
}

File: 31. Next Permutation/31. Next Permutation.rb
# @param {Integer[]} nums
# @return {Void} Do not return anything, modify nums in-place instead.
def next_permutation(nums)
    i = nums.length - 2
    while i >= 0 && nums[i] >= nums[i + 1]
        i -= 1
    end
    
    if i >= 0
        j = nums.length - 1
        while j >= 0 && nums[j] <= nums[i]
            j -= 1
        end
        nums[i], nums[j] = nums[j], nums[i]
    end
    
    reverse(nums, i + 1, nums.length - 1)
end

def reverse(nums, start, end_index)
    while start < end_index
        nums[start], nums[end_index] = nums[end_index], nums[start]
        start += 1
        end_index -= 1
    end
end